Improve slice browsing performance from 10fps to 90fps

lassoan opened the pull request:
Problem: On Intel i7 computers with NVidia GPUs, slice browsing was very slow (10-15fps).

Analysis: If "Threaded optimization" option in NVidia settings was disabled (by default it is enabled) then frame rate went up to 60fps (limited by vsync) to 89fps (without vsync).

Solution: Force image reslicing & mapping filters to use a single thread. It makes slice browsing fast (60-90fps) without disabling "Threaded optimization".

lassoan commented:
I have tested this on a desktop computer, where the dramatic speed improvement occurs. On a laptop computer without NVidia card, there seemed to be no performance difference.

Some testing can be done even without integrating these changes, just by forcing all filters to be single-threaded:

    vtk.vtkMultiThreader.SetGlobalMaximumNumberOfThreads(1)

jcfr commented:
I am curious to see what is the impact on the ScheduleRender() call with and without threading enabled.

It is interesting that we must disable threading on a CPU bound pipeline impact the GPU based rendering.

pieper commented:
On the mac, the current nightly is noticeably slower than 4.8.1 when interacting.  With the Reslicing button in the PerformanceTests module 4.8.1 is 96fps while the Nightly is 29fps.  

Setting vtk.vtkMultiThreader.SetGlobalMaximumNumberOfThreads(1) has no effect on either the feel of the application or the timings.

pieper commented:
This would be some logic to experiment with:

https://github.com/commontk/CTK/blob/fd7a896bd661fca2e885748869dcac3f47b5caba/Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp#L175-L246

lassoan commented:
> This would be some logic to experiment with:
https://github.com/commontk/CTK/blob/fd7a896bd661fca2e885748869dcac3f47b5caba/Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp#L175-L246

I've started my slice browsing slowdown investigation with exactly this part. I've found that this mechanism is not used for slice views in Slicer, as refresh rate is left at default 0.00001 FPS, which CTK interprets as "maximum FPS" (forceRender timer is set up with 0ms delay). This should only cause problems if there are extra application processEvents calls.

Anyway, I've fixed this mechanism and made maximum update rate configurable through settings. It is available here:
https://github.com/lassoan/Slicer/tree/limit-slice-refresh-rate

When refresh rate is set to a low enough value then Performance Tests module reports very high FPS, since many renderRequest calls are skipped. However, for me it only resulted in very small improvement in slice browsing experience (the slice slider or crosshair does not fall behind that much, but the its motion is not that continuous).

It would be great if you could test this with various desired frame rate values. If you use application settings to change default refresh rate then you need to restart Slicer. With this script you can adjust it without restarting the application:

```
red = slicer.mrmlScene.GetNodeByID('vtkMRMLSliceNodeRed')
yellow = slicer.mrmlScene.GetNodeByID('vtkMRMLSliceNodeYellow')
green = slicer.mrmlScene.GetNodeByID('vtkMRMLSliceNodeGreen')
rate=15 # FPS
for slice in [red,yellow,green]:
  slice.SetDesiredRefreshRate(rate/10.) # divide by 10 to compensate for CTK bug
```

Note that there is a 10x multiplier in the frame rate due to a CTK bug. This line https://github.com/commontk/CTK/blob/fd7a896bd661fca2e885748869dcac3f47b5caba/Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp#L188 should be ```msecsBeforeRender = 1000. / d->RenderWindow->GetDesiredUpdateRate()```.

lassoan commented:
I've also experimented with improving slice processing pipeline. Combination of various masking and lookup filters looked terribly inefficient (by reviewing the code). So, while profilers did not indicate them as bottlenecks, we reworked that part (with a help of a lab member, Thomas Vaughan). We replaced multiple filters with one, which does most processing in-place. See the branch here:

https://github.com/lassoan/Slicer/tree/slice-view-pipeline-optimization

lassoan commented:
> Neither solution you mentioned [disabling Threaded Optimization or using single thread] "feels" like the right solution to me. My gut thinks the fix is somewhere else.

I agree that disabling threaded optimization is not a clean solution. However, setting the filters to single-threaded feels right for me:

1. This is a 2D image processing pipeline, which processes 200-500x less amount of data than a volume processing pipeline. For a volume of 500 slices, using 8 cores, we would use 0.016 threads per slice. With default threading options, **for a single-slice volume we would use 8.0 threads per slice instead of a usual 0.016 threads per slice**.

2. To achieve the same performance (90fps) with default multithreading settings, **you would need to create and destroy 11340 threads per second** (for 8 core: 7 new threads per filter x 6 filters x 3 slice views x 90 fps). Creating this many threads is a significant overhead. The problem is made even worse by the fact that NVidia's "threaded optimization" uses heuristics to figure out what it should do in the main thread and what it should off-load to background threads. If we set number of threads to 1, then all processing is done in the main thread, therefore 0 new threads are created.

> We didn’t change anything in how image filters are parallelized. They use system threads libraries, pthreads on Linux and Mac.

Yes, this is not new. This terrible inefficiency has been haunting us for years. We did not know why strong computers have so poor reslicing performance. We discovered this now because with latest VTK version for some reason the frame rate dropped from 15fps (quite slow) to 8fps (unacceptably slow) and we had to do something.

Interestingly, on a Surface Book with Intel HD graphics 520 and NVidia GeForce GPU and HiDPI screen (3000x2000) is acting somewhat differently. There slice browsing is around 20fps with Intel HD and about 10fps with NVidia. Forcing single-threading and [simplifying filter pipeline](https://github.com/lassoan/Slicer/tree/slice-view-pipeline-optimization) increases speed by a couple of fps.

lassoan commented:
@pieper Could you please try what you get with https://github.com/lassoan/Slicer/tree/slice-view-pipeline-optimization + forcing single-threading? On my Surface book: using Intel HD I can achieve similar performance (about 10% slower) as with 4.8; using NVidia, the performance is even better than 4.8.

lassoan commented:
> On the mac, the current nightly is noticeably slower than 4.8.1 when interacting. With the Reslicing button in the PerformanceTests module 4.8.1 is 96fps while the Nightly is 29fps.

@pieper What Mac do you use? On a 2017 MacBook Air with Intel HD graphics 6000, we see 60fps (probably capped by vsync) with both 4.8.1 and latest nightly.

jcfr commented:
Cc: @robertmaynard

pieper commented:
> @pieper What Mac do you use? On a 2017 MacBook Air with Intel HD graphics 6000, we see 60fps (probably capped by vsync) with both 4.8.1 and latest nightly.

I tested with a mac pro with the firepro graphics (the version from when the mac pro first came out - I can check the exact version when I'm back in front of that computer in case that info would help).

lassoan commented:
Thanks for the advice @robertmaynard.

We can certainly crank up MinimumPieceSize, which will essentially lead to the same result as setting maximum thread count, but maybe a bit more elegant, as it may create more threads for unexpectedly large inputs.

We were aware of TBB and advantage of thread pools in theory, but we did not realize that we needed them. I'm still not sure that for special cases like this TBB is better than simply forcing single thread, but I'll build Slicer with TBB as VTK SMP backend and let you know how it works.

robertmaynard commented:
@lassoan Thanks for trying.

> We were aware of TBB and advantage of thread pools in theory, but we did not realize that we needed them. 

This happens to all of us :) . You only start to realize the cost of temporary thread construction after it has bitten you. 

> I'll build Slicer with TBB as VTK SMP backend and let you know how it works.

Thanks! I am crossing my fingers that my presumption on reducing the amount of thread creation will solve some of Slicers performance problems with slicing.

robertmaynard commented:
> Is there advantages in downloading the archives from Paraview staging area

The advantage is that we haven't tried the latest release of TBB and verified that the build system finds it correctly.

lassoan commented:
I've tried MinimumPieceSize but it did not improve the performance. There seems to be a fundamental issue with implementation of vtkThreadedImageAlgorithm::SplitExtent: it computes number of threads that are needed _after_ maximum number of threads have been already created. If the performance bottleneck is thread creation, none of the split strategies or piece sizes make any difference . This is a very basic flaw that should be relatively easy to fix and probably it would improve VTK's performance on small data sets (where processing time is negligible compared to thread creation time).

robertmaynard commented:
@lassoan Did you try TBB or are you still using the vtkMultiThreader?

lassoan commented:
Yes, I've tried TBB, I'm just writing a summary.

robertmaynard commented:
I agree that vtkThreadedImageAlgorithm once it detects it doesn't need all threads should lower the number of requested threads to vtkMultiThreader.

lassoan commented:
I've built Slicer using TBB as VTK SMP backend. Here is what I've found:
- Reslice of image data only: forcing single-threading => 96 fps, TBB => 130 fps.
- Reslice of image data in multiple views: forcing single-threading => 56 fps, TBB => 75 fps.
- Reslice of image+polydata: forcing single threading for image reslice only => 61 fps, forcing single-threading globally => 71 fps, TBB => 87 fps.
- Flying edge isosurfacing filter (only available using vtkSMPTools threading) is faster than marching cubes, for example on a typical data set marching cubes took 4.0sec while flying edge 2.3sec. This would be quite significant. Unfortunately, raw isosurfaces are rarely usable and if we apply surface smoothing filter as well, then the total execution time is about the same (8.3sec vs 7.9sec).

Conclusion:

1. For a real-time application, such as 3D Slicer, where most processing happens very quickly (so that results are visible in real-time), **globally forcing single-threading on an 8-core machine significantly improves performance when default Sequential SMP backend is used**. The problem is that currently VTK creates many unnecessary threads due to too small default piece size and a problem in multithreader (it always creates max number of threads, even if only 1 is used).

VTK's performance with Sequential SMP backend could be greatly improved by:
- fixing multithreader logic to only create threads that are actually used
- allowing global scaling of piece size (so that a balance could be found based on cost of thread creation and processing speed)
- setting sensible piece size for certain filters (e.g., those filters that perform trivial operations should probably have larger piece size, since computation time is comparable to thread creation time)

2. TBB significantly improves performance compared to Sequential SMP backend (even if Sequential backend thread counts optimized a bit manually).

**For Slicer, I think we should try to switch to TBB.** The Sequential backend seems to be abandoned by serious users (otherwise such fundamental flaw as unnecessary thread creation couldn't have been unnoticed for such a long time).

pieper commented:
Awesome research Andras!  I see that TBB is now licensed under Apache 2 rather than GPL, so that's great.

jcfr commented:
That is great news. Outstanding summary :+1: 

After talking with @wschroed , we wondered if all profiling was done in Release or RelWidthDebInfo ? I am assuming the answer is yes 

Also could you reference the fast marching and isosurface filters currently used ?

lassoan commented:
I used Release mode builds for getting the numbers above.

For isosurface generation I used vtkDiscreteMarchingCubes/vtkDiscreteFlyingEdges3D, optionally followed by vtkWindowedSincPolyDataFilter and vtkPolyDataNormals.

robertmaynard commented:
As piepe and jcfr have stated, outstanding benchmarking

You should be able to get both `vtkDiscreteMarchingCubes` and `vtkDiscreteFlyingEdges3D` to compute normals at the same time by enabling `SetComputeNormals`.

wschroed commented:
DiscreteMarchingCubes does not compute normals because it generates shared
vertices (e.g., shared between segmented regions). This vertex sharing
creates problems during vtkPolyDataNormals as well.

vtkDiscreteFlyingEdges3D creates separate regions for each contour value.
So it creates normals plus smoothing is better behaved. However it can
create coincident vertices, and after smoothing the regions can "pull away"
or intersect each other.



On Thu, Apr 5, 2018 at 5:41 PM, Robert Maynard <notifications@github.com>
wrote:

> As piepe and jcfr have stated, outstanding benchmarking
>
> You should be able to get both vtkDiscreteMarchingCubes and
> vtkDiscreteFlyingEdges3D to compute normals at the same time by enabling
> SetComputeNormals.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Slicer/Slicer/pull/930#issuecomment-379085585>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMMcvwNBCRtHfz1Jt0SNg6XTySDwjG7ks5tlo-GgaJpZM4TFo9B>
> .
>



-- 
William J. Schroeder, PhD
Kitware, Inc. - Building the World's Technical Computing Software
28 Corporate Drive
Clifton Park, NY 12065
will.schroeder@kitware.com
http://www.kitware.com
(518) 881-4902

lassoan commented:
I've tried normal computation of vtkDiscreteFlyingEdges3D and then apply smoothing, but the result was not smooth.

vtkDiscreteFlyingEdges3D (normals off) -> vtkWindowedSincPolyDataFilter -> vtkPolyDataNormal:

![image](https://user-images.githubusercontent.com/307929/38399540-00388f88-3919-11e8-8a67-5f2d8a277286.png)

vtkDiscreteFlyingEdges3D (normals on) -> vtkWindowedSincPolyDataFilter:

![image](https://user-images.githubusercontent.com/307929/38399559-246b7aaa-3919-11e8-89bc-e25da0bb7d7d.png)

lassoan commented:
@jcfr, do you think we should try to switch to TBB in Slicer-4.10? Do you need help with adding TBB build to CMake files?

jcfr commented:
>  do you think we should try to switch to TBB in Slicer-4.10?

After trying to build and package TBB on all the platforms, we should be able to conclude.

@robertmaynard  Is TBB currently packaged and distributed on all platforms with Paraview ? 

> Do you need help with adding TBB build to CMake files?

Yes. That would be great to get some help. I will not have a lot of time to allocate for this for the coming 2 or 3 weeks ... a lot of traveling.

lassoan commented:
@jcfr I've updated Slicer build so that you can specify SMP backend (default is still Sequential but you can set TBB).

https://github.com/Slicer/Slicer/pull/931

I have tested it on Windows but you would need to review the changes and fix up Linux and Mac implementations (I haven't done anything with rpaths and I was not sure which files need to be installed).

robertmaynard commented:
> Is TBB currently packaged and distributed on all platforms with Paraview ?c

@jcfr Yes ParaView currently packages and distributes TBB across all 3 major platforms.

lassoan commented:
Is there any potential drawback of using TBB over Sequential? Have you considered bundling TBB as third-party library to VTK (and maybe also enable it by default)?

robertmaynard commented:
> Is there any potential drawback of using TBB over Sequential? Maybe make TBB the default?

The imaging algorithms in VTK had already being threaded before the introduction of the SMP framework. The majority of the other algorithms had been designed for execution in a distributed-memory parallel model ( MPI ) where the larger application was assigning a single MPI rank per core.

The drawback of using TBB instead of sequential is that anyone executing inside this MPI model would over allocate as each MPI rank would spawn N workers. This is one of the reasons why TBB has not become the default for VTK. A longer term look at HPC environments is needed so that when launched with MPI, VTK is properly informed about how many threads to launch, what strategy to launch them in master/close/spread, how to pin them, and other HPC threading concerns.

So due to these concerns we haven't made TBB default

> Have you considered bundling TBB as third-party library to VTK

As far as I am aware the TBB build on all major platforms is a fairly opaque thing, and we would need to put serious effort in getting it converted to CMake, and build able on all three platforms. With the official binaries being sufficient for users, and HPC machines providing a TBB module it has made the need for a TBB third-party library pretty low.

lassoan commented:
Superseded by #931
